### ISSUE TITLE
Bug:  [Factories] preferApp does not work for classes with the same short name

### ISSUE THREAD

**Issue by kenjis** [codeigniter4/CodeIgniter4#7699]
Related: #7694

When we load a class in non-App namespace (`'Acme\Blog\Models\UserModel'`) with `'preferApp' => false`, 
and load a class which has the same short name (`'UserModel'` or `'App\Models\UserModel'`) with `'preferApp' => true`,  
the instance in non-App namespace will be returned.

> preferApp | boolean | Whether a class with the same basename in the App namespace overrides other explicit class requests.
https://codeigniter4.github.io/CodeIgniter4/concepts/factories.html#factories-options

### Steps to Reproduce

```diff
--- a/app/Config/Autoload.php
+++ b/app/Config/Autoload.php
@@ -42,6 +42,7 @@ class Autoload extends AutoloadConfig
     public $psr4 = [
         APP_NAMESPACE => APPPATH, // For custom app namespace
         'Config'      => APPPATH . 'Config',
+        'Acme\Blog'   => ROOTPATH . 'acme/Blog',
     ];
 
     /**
```
```php
<?php

namespace App\Models;

use CodeIgniter\Model;

class UserModel extends Model
{
    protected $table = 'users';
}
```
```php
<?php

namespace Acme\Blog\Models;

class UserModel
{
}
```
```php
<?php

namespace App\Controllers;

use CodeIgniter\Config\Factories;

class Home extends BaseController
{
    public function index()
    {
        $users1 = Factories::models('Acme\Blog\Models\UserModel', ['preferApp' => false]);
        $users2 = Factories::models('UserModel', ['preferApp' => true]);

        return get_class($users1) . ', ' . get_class($users2);
    }
}
```
Results:
```
Acme\Blog\Models\UserModel, Acme\Blog\Models\UserModel
```

**Comment by kenjis** [comment]
Although difficult to understand, this may not be a bug.

Once an instance is held in Factories, it is a specification that the shared instance will be returned.

When `'UserModel'` is specified, the shared `UserModel` instance is returned, so the option `'preferApp' => true` passed at runtime is ignored.

**Comment by kenjis** [comment]
Closed by v4.4.0.
